### PR TITLE
Allow wildcards for java class lookup transform providers.

### PR DESCRIPTION
**Pull request by robertwb**
This is especially useful for cases where the expansion service is implicitly
started by the user for ad-hoc invocation of cross-language transforms (in which
case there is no security benefit to an explicit allow list).

Also automatically starts up an expansion service for Python's
JavaExternalTransform and adds the ability to manually specify a classpath.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

`ValidatesRunner` compliance status (on master branch)
--------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>Lang</th>
      <th>ULR</th>
      <th>Dataflow</th>
      <th>Flink</th>
      <th>Samza</th>
      <th>Spark</th>
      <th>Twister2</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Go</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/">
          Build Status
        </a>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Samza/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
    </tr>
    <tr>
      <td>Java</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_ULR/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Streaming/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_VR_Dataflow_V2/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_VR_Dataflow_V2_Streaming/lastCompletedBuild/">
          Build Status
        </a><br>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Samza/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/">
          Build Status
        </a>
      </td>
    </tr>
    <tr>
      <td>Python</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Samza/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
    </tr>
    <tr>
      <td>XLang</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Dataflow/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Samza/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
    </tr>
  </tbody>
</table>

Examples testing status on various runners
--------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>Lang</th>
      <th>ULR</th>
      <th>Dataflow</th>
      <th>Flink</th>
      <th>Samza</th>
      <th>Spark</th>
      <th>Twister2</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Go</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
    <tr>
      <td>Java</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Java_Examples_Dataflow_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Java_Examples_Dataflow_Java11_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_Examples_Dataflow_V2/lastCompletedBuild/">
          Build Status
        </a><br>
      </td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
    <tr>
      <td>Python</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
    <tr>
      <td>XLang</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
  </tbody>
</table>

Post-Commit SDK/Transform Integration Tests Status (on master branch)
------------------------------------------------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>Go</th>
      <th>Java</th>
      <th>Python</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python38/lastCompletedBuild/">
          Build Status
        </a>
      </td>
    </tr>
  </tbody>
</table>

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>---</th>
      <th>Java</th>
      <th>Python</th>
      <th>Go</th>
      <th>Website</th>
      <th>Whitespace</th>
      <th>Typescript</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Non-portable</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_PythonDocker_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_PythonDocs_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Whitespace_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Typescript_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
    </tr>
    <tr>
      <td>Portable</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_GoPortable_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
  </tbody>
</table>

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by robertwb** [comment]
R: @lukecwik , @chamikaramj

File: sdks/python/apache_beam/transforms/external.py [hunk 1]
```diff
@@ -283,9 +283,21 @@ def _has_constructor(self):
...
 
-  One builds these transforms just as one would in Java.
+  One builds these transforms just as one would in Java, e.g.::
+
+      transform = JavaExternalTransform('fully.qualified.ClassName')(
+          contrustorArg, ...).builderMethod(...)
+
+  or::
+
+      JavaExternalTransform('fully.qualified.ClassName').staticConstructor(
```
Comment (chamikaramj) [inline sdks/python/apache_beam/transforms/external.py:293]: JavaExternalTransform('fully.qualified.ClassName')(\<params\>) works as well, right ?

File: sdks/python/apache_beam/transforms/external.py [hunk 2]
```diff
@@ -283,9 +283,21 @@ def _has_constructor(self):
...
+          ...).builderMethod1(...).builderMethod2(...)
   """
-  def __init__(self, class_name, expansion_service=None):
+  def __init__(self, class_name, expansion_service=None, classpath=None):
+    expansion_service = expansion_service or BeamJarExpansionService(
```
Comment (chamikaramj) [inline sdks/python/apache_beam/transforms/external.py:297]: So the default version will just offer the classes packaged with the <expansion-service> package ? I'm not sure whether this will be useful (it's whatever in that package and Beam core). Should we make providing classpath (or additional jars) required to make the default usage useful ?

File: sdks/python/apache_beam/transforms/external.py [hunk 3]
```diff
@@ -655,11 +667,13 @@ class JavaJarExpansionService(object):
   argument which will spawn a subprocess using this jar to expand the
   transform.
   """
-  def __init__(self, path_to_jar, extra_args=None):
+  def __init__(self, path_to_jar, extra_args=None, classpath=None):
     if extra_args is None:
-      extra_args = ['{{PORT}}', f'--filesToStage={path_to_jar}']
+      to_stage = ','.join([path_to_jar] + list(classpath or []))
```
Comment (chamikaramj) [inline sdks/python/apache_beam/transforms/external.py:672]: Seems like we assume that the provided classpath only contains jars ? Classpath also supports directories and wild cards.
https://docs.oracle.com/javase/8/docs/technotes/tools/windows/classpath.html

File: sdks/python/apache_beam/transforms/external.py [hunk 4]
```diff
@@ -283,9 +283,21 @@ def _has_constructor(self):
...
+      transform = JavaExternalTransform('fully.qualified.ClassName')(
+          contrustorArg, ...).builderMethod(...)
+
+  or::
+
+      JavaExternalTransform('fully.qualified.ClassName').staticConstructor(
+          ...).builderMethod1(...).builderMethod2(...)
   """
-  def __init__(self, class_name, expansion_service=None):
+  def __init__(self, class_name, expansion_service=None, classpath=None):
```
Comment (chamikaramj) [inline sdks/python/apache_beam/transforms/external.py:296]: Please document what classpath should be (seems like a list based on the usage below).

File: sdks/python/apache_beam/utils/subprocess_server.py [hunk 5]
```diff
@@ -158,9 +158,14 @@ class JavaJarServer(SubprocessServer):
       'local', (threading.local, ),
       dict(__init__=lambda self: setattr(self, 'replacements', {})))()
 
-  def __init__(self, stub_class, path_to_jar, java_arguments):
+  def __init__(self, stub_class, path_to_jar, java_arguments, classpath=None):
```
Comment (chamikaramj) [inline sdks/python/apache_beam/utils/subprocess_server.py:161]: Could we add a unit test to cover this ?

**Review by chamikaramj — COMMENTED** [review]
Thanks!

File: sdks/java/expansion-service/src/main/java/org/apache/beam/sdk/expansion/service/JavaClassLookupTransformProvider.java [hunk 6]
```diff
@@ -508,12 +523,32 @@ static AllowList create(
...
+    public static final List<String> WILDCARD = Collections.singletonList("*");
+
     public abstract String getClassName();
 
     public abstract List<String> getAllowedBuilderMethods();
 
     public abstract List<String> getAllowedConstructorMethods();
 
+    public boolean isAllowedClass(String className) {
+      String pattern = getClassName();
```
[the remark this answers is not in the document]
Comment (robertwb) [reply sdks/java/expansion-service/src/main/java/org/apache/beam/sdk/expansion/service/JavaClassLookupTransformProvider.java:535]: I could find documentation of this file anywhere, but I think the code is self explanatory enough (for both non-wildcard and wild-card containing lists).

File: sdks/python/apache_beam/utils/subprocess_server.py (hunk 5 above)
Comment (robertwb) [reply sdks/python/apache_beam/utils/subprocess_server.py:161]: Done.

**Comment by robertwb** [comment]
Also rebased on top of https://github.com/apache/beam/pull/15891

File: sdks/python/apache_beam/transforms/external.py [hunk 7]
```diff
@@ -655,22 +680,29 @@ class JavaJarExpansionService(object):
   argument which will spawn a subprocess using this jar to expand the
   transform.
   """
-  def __init__(self, path_to_jar, extra_args=None):
-    if extra_args is None:
```
Comment (chamikaramj) [inline sdks/python/apache_beam/transforms/external.py:659]: Probably need to rebase again (based on the diff).

File: sdks/python/apache_beam/transforms/external.py [hunk 8]
```diff
@@ -655,22 +680,29 @@ class JavaJarExpansionService(object):
...
-      extra_args = ['{{PORT}}', f'--filesToStage={path_to_jar}']
+  def __init__(self, path_to_jar, extra_args=None, classpath=None):
     self._path_to_jar = path_to_jar
     self._extra_args = extra_args
+    self._classpath = classpath
     self._service_count = 0
 
+  def _default_args(self):
+    to_stage = ','.join([self._path_to_jar] + sum(
```
Comment (chamikaramj) [inline sdks/python/apache_beam/transforms/external.py:690]: Fail here if self._path_to_jar is not set ?

File: sdks/python/apache_beam/transforms/external_test.py [hunk 9]
```diff
@@ -547,6 +550,27 @@ def test_implicit_builder_with_constructor_method(self):
     self._verify_row(build_method.schema, build_method.payload, {})
 
 
+class JavaJarExpansionServiceTest(unittest.TestCase):
+  def test_classpath(self):
```
Comment (chamikaramj) [inline sdks/python/apache_beam/transforms/external_test.py:554]: BTW I think we should add more comprehensive unit testing for "JavaJarExpansionService" and "BeamJarExpansionService" (can be a separate PR).

File: sdks/python/apache_beam/utils/subprocess_server.py [hunk 10]
```diff
@@ -158,9 +158,14 @@ class JavaJarServer(SubprocessServer):
...
+    if classpath:
+      cp_args = ['-classpath', os.pathsep.join(classpath)]
+    else:
+      cp_args = []
     super().__init__(
-        stub_class, ['java', '-jar', path_to_jar] + list(java_arguments))
+        stub_class,
+        ['java'] + cp_args + ['-jar', path_to_jar] + list(java_arguments))
```
Comment (chamikaramj) [inline sdks/python/apache_beam/utils/subprocess_server.py:168]: Unfortunately I don't think "java -jar" command can take a CLASSPATH as an argument. So this probably won't work.
https://docs.oracle.com/en/java/javase/17/docs/specs/man/jar.html

Im fine with partially rolling this back to use a single jar when we automatically startup the expansion service.

**Review by chamikaramj — COMMENTED** [review]
Thanks.

File: sdks/python/apache_beam/utils/subprocess_server.py (hunk 10 above)
Comment (chamikaramj) [reply sdks/python/apache_beam/utils/subprocess_server.py:168]: This is the correct link: https://docs.oracle.com/javase/7/docs/technotes/tools/windows/java.html

"When you use this option, the JAR file is the source of all user classes, and other user class path settings are ignored."

File: sdks/python/apache_beam/transforms/external.py [hunk 11]
```diff
@@ -302,7 +318,11 @@ def __getattr__(self, name):
     # Don't try to emulate special methods.
     if name.startswith('__') and name.endswith('__'):
       return super().__getattr__(name)
+    else:
+      return self[name]
 
+  def __getitem__(self, name):
```
[the remark this answers is not in the document]
Comment (robertwb) [reply sdks/python/apache_beam/transforms/external.py:324]: It just makes it so we can use names that would otherwise overlap with existing attributes.

File: sdks/python/apache_beam/transforms/external_test.py (hunk 9 above)
Comment (robertwb) [reply sdks/python/apache_beam/transforms/external_test.py:554]: Yeah, I agree.

File: sdks/python/apache_beam/utils/subprocess_server.py (hunk 10 above)
Comment (robertwb) [reply sdks/python/apache_beam/utils/subprocess_server.py:168]: Oh, that's really unfortunate. This really limits the ability to use classes in an ad-hoc way... I'll have to think about this some more.

File: sdks/python/apache_beam/transforms/external.py (hunk 8 above)
Comment (robertwb) [reply sdks/python/apache_beam/transforms/external.py:690]: It will always be set. This is a private method, only factored out for testing.

File: sdks/python/apache_beam/utils/subprocess_server.py (hunk 10 above)
Comment (robertwb) [reply sdks/python/apache_beam/utils/subprocess_server.py]: OK, I added support for this. I was tempted to just drop the classpath thing, but then the default JavaExternalTransform would be much less useful, and I think this'll be needed for the JDBC transforms as well.

**Review by robertwb — COMMENTED** [review]
Thanks, PTAL.

File: sdks/python/apache_beam/utils/subprocess_server.py [hunk 12]
```diff
@@ -158,7 +161,11 @@ class JavaJarServer(SubprocessServer):
       'local', (threading.local, ),
       dict(__init__=lambda self: setattr(self, 'replacements', {})))()
 
-  def __init__(self, stub_class, path_to_jar, java_arguments):
+  def __init__(self, stub_class, path_to_jar, java_arguments, classpath=None):
+    if classpath or True:
```
Comment (chamikaramj) [inline sdks/python/apache_beam/utils/subprocess_server.py:165]: This is always True ?

File: sdks/python/apache_beam/utils/subprocess_server.py [hunk 13]
```diff
@@ -283,6 +290,39 @@ def beam_services(cls, replacements):
     finally:
       cls._BEAM_SERVICES.replacements = old
 
+  @classmethod
+  def make_classpath_jar(cls, main_jar, extra_jars, cache_dir=None):
```
Comment (chamikaramj) [inline sdks/python/apache_beam/utils/subprocess_server.py:294]: Please add a unit test for this as well.

File: sdks/python/apache_beam/utils/subprocess_server.py [hunk 14]
```diff
@@ -283,6 +290,39 @@ def beam_services(cls, replacements):
...
+    if cache_dir is None:
+      cache_dir = cls.JAR_CACHE
+    composite_jar_dir = os.path.join(cache_dir, 'composite-jars')
+    os.makedirs(composite_jar_dir, exist_ok=True)
+    classpath = []
+    # Class-Path references from a jar must be relative.
```
Comment (chamikaramj) [inline sdks/python/apache_beam/utils/subprocess_server.py:300]: Could you expand the comment here to explain what's done here.

File: sdks/python/apache_beam/utils/subprocess_server.py [hunk 15]
```diff
@@ -283,6 +290,39 @@ def beam_services(cls, replacements):
...
+        rel_path = hashlib.sha256(
+            path.encode('utf-8')).hexdigest() + os.path.splitext(path)[1]
+        classpath.append(rel_path)
+        if not os.path.lexists(os.path.join(composite_jar_dir, rel_path)):
+          os.symlink(path, os.path.join(composite_jar_dir, rel_path))
+    composite_jar = os.path.join(
+        composite_jar_dir,
+        hashlib.sha256(' '.join(sorted(classpath)).encode('ascii')).hexdigest()
+        + '.jar')
+    if not os.path.exists(composite_jar):
```
Comment (chamikaramj) [inline sdks/python/apache_beam/utils/subprocess_server.py:312]: It's unclear to me if this will work for all Jars. @lukecwik might be able to confirm. (non-blocking comment)

**Review by chamikaramj — COMMENTED** [review]
Thanks. LGTM other than few comments. 

Also, please consider adding an integration test to make sure that composite jar creation works end-to-end (separately).

**Comment by chamikaramj** [comment]
Hi Robert, any updates on this ?
It's great if we can get this into Beam 2.35.0 (cut 11/17)

File: sdks/python/apache_beam/utils/subprocess_server.py (hunk 13 above)
Comment (robertwb) [reply sdks/python/apache_beam/utils/subprocess_server.py:294]: Done.

**Comment by chamikaramj** [comment]
Seems like Python failure is related: https://ci-beam.apache.org/job/beam_PreCommit_Python_Commit/20540/testReport/junit/apache_beam.utils.subprocess_server_test/JavaJarServerTest/test_classpath_jar/

**Comment by robertwb** [comment]
Looks like someone is monkey-patching mock into there and it's not getting un-mocked... It passed locally. I'll try to fix this.

**Comment by robertwb** [comment]
Python is now happy (we now make sure we unbock in processes_test.py) and the java issues look unrelated.

**Comment by chamikaramj** [comment]
Thanks. LGTM

**Comment by lukecwik** [comment]
The change in the shadow closure added the JUL logging adapter creating an infinite loop when logging during spark and flink tests.

Rolling back.
